### PR TITLE
[SEDONA-428] Add RS_ZonalStats & RS_ZonalStatsAll

### PR DESCRIPTION
**Pull request by furqaankhan**
## Did you read the Contributor Guide?

- Yes, I have read [Contributor Rules](https://sedona.apache.org/latest-snapshot/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest-snapshot/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes, the URL of the associated JIRA ticket is https://issues.apache.org/jira/browse/SEDONA-428. The PR name follows the format `[SEDONA-XXX] my subject`.

## What changes were proposed in this PR?

- Add RS_ZonalStats
- Add RS_ZonalStatsAll
- Refactor asRaster implementation to have the ability to keep the reference raster's extent.

## How was this patch tested?

- Passed new and old tests

## Did this PR include necessary documentation updates?

- Yes, I have updated the documentation update.

### CODE REVIEW

**Review by jiayuasu — CHANGES_REQUESTED** [review]

**Comment by jiayuasu** [comment]
@Kontinuation All raster functions that take a raster and a geometry as an input, we must perform implicit CRS transformation. What do you think?

File: common/src/main/java/org/apache/sedona/common/raster/RasterConstructors.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
 import org.geotools.gce.geotiff.GeoTiffReader;
 import org.geotools.geometry.Envelope2D;
 import org.geotools.geometry.jts.JTS;
```
Comment (jiayuasu) [inline common/src/main/java/org/apache/sedona/common/raster/RasterConstructors.java:28]: Why does this PR involve changes related to AsRaster? Is this a API breaking change?

If this change is not related to zonal statistics. It must be placed in a separate PR.

File: common/src/main/java/org/apache/sedona/common/raster/RasterBandAccessors.java [hunk 2]
```diff
@@ -81,6 +88,190 @@ public static long getCount(GridCoverage2D raster, int band) {
...
+        // checking if the raster contains the geometry
+        if (!RasterPredicates.rsIntersects(raster, roi)) {
+            throw new IllegalArgumentException("The provided geometry is not intersecting the raster. Please provide a geometry that is in the raster's extent.");
+        }
+
+        Raster rasterData = RasterUtils.getRaster(raster.getRenderedImage());
+        String datatype = RasterBandAccessors.getBandType(raster, band);
+        Double noDataValue = RasterBandAccessors.getBandNoDataValue(raster, band);
+        // Adding an arbitrary value '150' for the pixels that are under the geometry.
+        GridCoverage2D rasterizedGeom = RasterConstructors.asRasterWithRasterExtent(roi, raster, datatype, 150, null);
```
Comment (jiayuasu) [inline common/src/main/java/org/apache/sedona/common/raster/RasterBandAccessors.java:243]: Why add `150` here?

File: common/src/test/java/org/apache/sedona/common/raster/RasterBandAccessorsTest.java [hunk 3]
```diff
@@ -75,6 +81,113 @@ public void testBandNoDataValueIllegalBand() throws FactoryException, IOExceptio
...
 
+    @Test
+    public void testZonalStats() throws FactoryException, ParseException, IOException {
+        GridCoverage2D raster = rasterFromGeoTiff(resourceFolder + "raster_geotiff_color/FAA_UTM18N_NAD83.tif");
+        String polygon = "POLYGON ((236722 4204770, 243900 4204770, 243900 4197590, 221170 4197590, 236722 4204770))";
+        Geometry geom = Constructors.geomFromWKT(polygon, RasterAccessors.srid(raster));
+
+        double actual = RasterBandAccessors.getZonalStats(raster, geom, 1, "sum", false);
+        double expected = 1.0690406E7;
+        assertEquals(expected, actual, 0d);
```
Comment (jiayuasu) [inline common/src/test/java/org/apache/sedona/common/raster/RasterBandAccessorsTest.java:92]: Please set the allowed delta value as a parameter in the raster test base such as ` protected static final double FP_TOLERANCE = 1e--4`
`

File: docs/api/sql/Raster-operators.md [hunk 4]
```diff
@@ -937,6 +937,129 @@ Output:
...
+- 4: Mode.
+- 5: Standard deviation.
+- 6: Variance.
+- 7: Minimum value of the zone.
+- 8: Maximum value of the zone.
+
+!!!note
+    The following conditions will throw an `IllegalArgumentException` if they are not met:
+
+    - The `raster` and `zone` geometry should be in the same CRS, refer to [RS_SRID](#rs_srid) for raster, and [ST_SRID](../Function/#st_srid) for geometry to get the spatial reference identifier.
```
Comment (jiayuasu) [inline docs/api/sql/Raster-operators.md:1018]: Please explain what happens if either side does not have CRS or different CRS.

**Review by jiayuasu — CHANGES_REQUESTED** [review]

File: common/src/main/java/org/apache/sedona/common/raster/RasterBandAccessors.java (hunk 2 above)
Comment (jiayuasu) [reply common/src/main/java/org/apache/sedona/common/raster/RasterBandAccessors.java:243]: @furqaankhan Will it collide with existing values in a raster? If not, we are good.

Comment (furqaankhan) [reply common/src/main/java/org/apache/sedona/common/raster/RasterBandAccessors.java:243]: No no. It just to determine which values to use for compute from the provided raster.

**Review by jiayuasu — APPROVED** [review]
